### DISCUSSION TITLE
Bulk-insert with multiple cascade, back_populate relationships

### DISCUSSION THREAD

**Discussion by AxelTob** [sqlalchemy/sqlalchemy#8154] [Usage Questions]
I have tried to optimize our insertions to the database, which is currently the bottleneck and slowing down our pipeline. I decided to first start to speed up our data_generator used for testing, all the tables are empty at first. Thought it would be an easy place to start .. they are then populated and used in various tests. Currently, we do pretty much all insertions with _Session.add(entry)_ or in some cases bulked entries with _add_all(entries)_, which does not improve the speed that much.

The goal was to do more insertions at once and have less time communicating back and forth with the database and I tried various bulk_insert methods _(bulk_save_objects, bulk_insert_mappings_, and ORM, CORE methods with _INSERT INTO, COPY, IMPORT_ .. but I _got nothing to work properly_. **Foreign key constraints, duplicated keys ..**. or tables not getting populated.

I will show an example of a Table that would previous be added with add_all() in a run_transaction.
```
class News(NewsBase):
__tablename__ = 'news'

news_id = Column(UUID(as_uuid=True), primary_key=True, nullable=False)
url_visit_count = Column('url_visit_count', Integer, default=0)

# One to many
sab_news = relationship("sab_news", back_populates="news")
sent_news = relationship("SenNews", back_populates="news")
scope_news = relationship("ScopeNews", back_populates="news")
news_content = relationship("NewsContent", back_populates="news")

# One to one
other_news = relationship("other_news", uselist=False, back_populates="news")

# Many to many
companies = relationship('CompanyNews', back_populates='news', cascade="all, delete")
aggregating_news_sources = relationship("AggregatingNewsSource", secondary=NewsAggregatingNewsSource,
                                        back_populates="news")

def __init__(self, title, language, news_url, publish_time):
    self.news_id = uuid4()
    super().__init__(title, language, news_url, publish_time)
```

We have many tables built like this, some with more relations, and my conclusion now is that having many different relationships that **back_populates and update each other does not allow for fast bulk_insertions, Am I wrong?**

One of my current solution that was able to decrease our execution_time from 120s to 15s for a regular data_generator for testing is like this:
```
def write_news_to_db(news, news_types, news_sources, company_news,
                 ):
write_bulk_in_chunks(news_types)
write_bulk_in_chunks(news_sources)

def write_news(session):
    enable_batch_inserting(session)
    session.add_all(news)

def write_company_news(session):
    session.add_all(company_news)

engine = create_engine(
        get_connection_string("name"),
        echo = False,
        executemany_mode = "values")

run_transaction(create_session(engine=engine), lambda s: write_news(s))
run_transaction(create_session(), lambda s: write_company_news(s))
```

I used this library [batch_insert](https://github.com/benchling/sqlalchemy_batch_inserts) together with Psycopg2 Fast Execution Helpers, set executemany_mode="values". _I did this by creating a new engine just for these insertions - It did work however this itself seems like a bad practice. It works with the same database._

Anyway, this does seem to work, but it is still not the execution speed I want - especially when we are initially working with empty tables.

Ideally, I wouldn't want to do this hacky solution and avoid bulk_insertions since SQLAlchemy does not recommend using them - to avoid problems that I have faced.

**But how does one construct queries to properly do bulk_insertions in cases of complex Tables like these - should we re-design our tables or is it possible?**

Using Multi-row insertions within the run_transaction with _ORM or CORE would be ideal,_ but I haven't been able to do it.

Any recommendations or help would be much appreciated!

**TLDR; Bulk-insertion with multiple relationships, back_populates, cascade. How is it supposed to done?**

**Comment by zzzeek** [comment]
back_populates doesn't really have much impact on insertion speed, except that maybe it is adding Python overhead when you are first constructing object structures.

with the psycopg2 driver, SQLAlchemy 1.4 should already be batching INSERT statements together into a single INSERT per table with one big VALUES, unless you have relationships that are self-referential.  If those classes up there like SenNews, ScopeNews are actually subclasses of News, then that's a self-referential relationship which might be resulting in INSERT statements more broken up.

anyway there are a lot of varieties of "execution speed" here and they depend a  lot on how many rows we are talking about.    PostgreSQL's bulk insert facility is the COPY statement, so if you are looking for "lightning fast", you'd be looking at that.

if this is a fixed data fixture for a test suite, take a [dump of the database](https://www.postgresql.org/docs/current/backup-dump.html) with all the data in it into a file that consists of COPY statements.  Then use that script to build up your fixtures.   That should be an order of magnitude faster than anything you can do in Python.

**Comment by gordthompson** [comment]
My 2¢:

Often when people want to use related ORM objects to "bulk insert" things it's because they're using autoincrement PKs and they want SQLAlchemy to take care of the foreign keys for them. (Insert parent, get parent PK, use parent PK as FK when inserting child.)

However, in your case you appear to be using UUID PKs and specifying them yourself (ref: `self.news_id = uuid4()`) so you're not relying on a server-side process to generate them for you. That leads me to expect that you would be able to structure your test data with all of the PKs and FKs known in advance and therefore you could use more low-level techniques to populate the tables and not incur the overhead of ORM and Session and unit-of-work and so on.

**Comment by gordthompson** [comment]
Note also that SQLAlchemy 1.4 by default works with psycopg2's Fast Execution Helpers in [values_only](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html#psycopg2-fast-execution-helpers) mode so multiple inserts are "batched" in the VALUES clause without the need for the [batch_inserts](https://github.com/benchling/sqlalchemy_batch_inserts) add-on (or similar). Example:

```python
class Thing(Base):
    __tablename__ = "thing"
    id = Column(Integer, primary_key=True)
    txt = Column(String(50))


Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

with Session(engine) as sess:
    sess.add_all(
        [
            Thing(txt="foo"),
            Thing(txt="bar"),
            Thing(txt="baz"),
        ]
    )
    sess.commit()
```

sends this over the wire …

```
0000   00 21 70 75 83 c1 40 2c f4 34 c0 d3 08 00 45 00   .!pu..@,.4....E.
0010   00 77 1d 6b 40 00 80 06 5a 4b c0 a8 00 b3 c0 a8   .w.k@...ZK......
0020   00 c7 c3 47 66 91 fd a1 06 d4 c0 4a f8 42 50 18   ...Gf......J.BP.
0030   00 fb 63 bf 00 00 51 00 00 00 4e 49 4e 53 45 52   ..c...Q...NINSER
0040   54 20 49 4e 54 4f 20 74 68 69 6e 67 20 28 74 78   T INTO thing (tx
0050   74 29 20 56 41 4c 55 45 53 20 28 27 66 6f 6f 27   t) VALUES ('foo'
0060   29 2c 28 27 62 61 72 27 29 2c 28 27 62 61 7a 27   ),('bar'),('baz'
0070   29 20 52 45 54 55 52 4e 49 4e 47 20 74 68 69 6e   ) RETURNING thin
0080   67 2e 69 64 00                                    g.id.
```

which we can see includes

```sql
INSERT INTO thing (txt) VALUES ('foo'),('bar'),('baz') RETURNING thing.id
```
